### ISSUE TITLE
can't use affnity

### ISSUE THREAD

**Issue by Readm** [herd/herdtools7#410]
I tried to run litmus7 with -p option, but some case didn't work (some did, same enviorment).
I also tried to remove all preamble, only left the init part and test part. Still not working. Does the affinity relies on anything I missed? Thanks.

**Comment by maranget** [comment]
Hi @Readm, could you please describe your problem in greater detail: (1) system used (architecture, number of cores ofd target machine, OS, litmus version...), (2) exact command line (including litmus test), (3) expected and actual behaviour.

Thanks.

**Comment by Readm** [comment]
system: M1 chip (aarch64), 8 cores, Asahi Linux
command line: `litmus7 MP.litmus -o xxx.tar`, then `tar xf xxx.tar` and `make`
If it works the ELF file shuold support `./MP.exe -p n,m`
But it shows the Error and shows the usage, and in the usage, there is no `-p` option (if it works, `-p` is in the help info).

Litmus test:
```
AArch64 MP
"PodWW Rfe PodRR Fre"

Cycle=Rfe PodRR Fre PodWW
Relax=
Safe=Rfe Fre PodWW PodRR
Prefetch=0:x=F,0:y=W,1:y=F,1:x=T
Com=Rf Fr
Orig=PodWW Rfe PodRR Fre

{
0:X1=x; 0:X3=y;
1:X1=y; 1:X3=x;
}

 P0                      | P1                      ;
 MOV W0,#1               | LDR W0,[X1]  (* Ry=1 *) ;
 STR W0,[X1]  (* Wx=1 *) | LDR W2,[X3]  (* Rx=0 *) ;
 MOV W2,#1               |                         ;
 STR W2,[X3]  (* Wy=1 *) |                         ;

exists
(1:X0=1 /\ 1:X2=0)

```
But this one worked:
```
AArch64 2+2W
"PodWW Wse PodWW Wse"
Cycle=Wse PodWW Wse PodWW
Relax=
Safe=Wse PodWW
Prefetch=0:x=F,0:y=W,1:y=F,1:x=W
Com=Ws Ws
Orig=PodWW Wse PodWW Wse
{
0:X1=x; 0:X3=y;
1:X1=y; 1:X3=x;
}
 P0          | P1          ;
 MOV W0,#2   | MOV W0,#2   ;
 STR W0,[X1] | STR W0,[X1] ;
 MOV W2,#1   | MOV W2,#1   ;
 STR W2,[X3] | STR W2,[X3] ;
exists
(x=2 /\ y=2)

```

Thanks

**Comment by Readm** [comment]
Plus, If I do not use `-o xxx.tar` and use litmus7 to run it with `-p`, it does not report error, but in the running details, the affinity shows the affinity is `None`.

Sorry the test machine is not convient to copy log here.

**Comment by maranget** [comment]
Your command line shows no option. Hence, affinity control is disabled:
```
% litmus7 -help 2>&1 | grep affinity
  -affinity <none|random|custom|incr<n>|scan> attach threads to logical processors, default none
```
In other words, the behaviour you observe for test `MP` is normal.

 I am surprised you have affinity enabled for the other test (see below). Namely, litmus7 on MacOS does not support affinity control. As far as I know, MacOS provides no API for this. As a consequence, if you attempt to enable affinity control (_e.g._ `litmus7 -i 1...`), gcc compilation will fail.

There are some configuration files in litmus distribution. In particular, there is one for M1 computers:
```
% litmus7 -mach M1 -i 1 -o R.tar MP.litmus
Fatal error: exception Misc.Fatal("Affinity not implemented for mac")
```

**Comment by Readm** [comment]
Thanks, it is not MacOS, it is [Asahi Linux](https://asahilinux.org/about/), a Linux for M1. Maybe that's why it works?
So the rule is, if I want to use affinity control by output to a tar file and make, I must add affinity option when I generate the tar file? 
But in which I said working test, affinity can be re-configured even after I compile the ELF file. Will this overwrite the opition when calling litmus7?

**Comment by Readm** [comment]
I tried to add `litmus7 MP.litmus -p 6,7 -o xxx.tar` . Still not working, really strange that another one works. It seems the option is ignored, it didn't raise an Error while generating tar.

**Comment by maranget** [comment]
Yes you  must add some affinity control option to litmus so as to enable affinity control, you can then modify the affinity options with `.exe` command line options. There may be a glitch: documentation states that option -p implies option `-affinity incr1` but it does not seem so. May I suggest that you try the following command line:
```
% mkdir R
% litmus -p 6,7 -i 1 -o R MP.litmus
% cd R
% make
% ./MP.exe -v
```
You should see something like:
```
MP: n=1, r=10, s=100000, i=1, p='6,7'
...
...

**Comment by Readm** [comment]
Thank a lot!!! It works.
So maybe in previous generation I added -i 1, but I forgot that. I did think that it is default thus not usefull.

**Comment by maranget** [comment]
Glitch fixed, PR #412
